### ISSUE TITLE
Nested queries do not work with H2

### ISSUE THREAD

**Issue by dya-tel** [JetBrains/Exposed#131]
Exposed 0.8.1
H2 1.4.196, embedded

When database is queried, any existing queries are being closed for some reason, so if someone makes a query while processing another query in the same transaction, application crashes.
This behavior is not consistent with MySQL driver, so it is probably a bug.

`Table1.selectAll().map { Table2.selectAll().filter { true }; it }`

21:09:42.878 [Test worker] INFO  Exposed - Transaction attempt #0: The object is already closed [90007-196]. Statement: null
org.h2.jdbc.JdbcSQLException: The object is already closed [90007-196]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.message.DbException.get(DbException.java:144)
	at org.h2.jdbc.JdbcResultSet.checkClosed(JdbcResultSet.java:3208)
	at org.h2.jdbc.JdbcResultSet.next(JdbcResultSet.java:130)
	at org.jetbrains.exposed.sql.Query$ResultIterator.hasNext(Query.kt:250)

**Comment by Tapac** [comment]
H2 (beside most of other databases) doesn't support concurrent resultsets within single connection.
You should "read" whole result before iteration. 
Try ```Table1.selectAll().toList().map { Table2.selectAll().filter { true }; it } ```

**Comment by dya-tel** [comment]
This solution works, but it kind of makes API inconsitent.
As a user, I expect that if I got a resultset, the query is already done and I can do anything I want. And I already get this behavior with every database, but not with H2.
And so it would be very nice if it will be handled inside the library. If fixing it would decrease performance, "internal result reading" could be applied only when H2 dialect is being used.

There is also a bug (?) with SQLite: it doesn't support current default transaction isolation level. The library should set a different default value when used with SQLite (probably, TRANSACTION_SERIALIZABLE). Manual setting of this value fixes it, but it brings API inconsistency again.

**Comment by Tapac** [comment]
selectAll(), select and other Query functions return SizedIterable - lazy iterator over ResultSet. This approach allows you control amount of data you want to retrieve from database. Eager fetching may result in OOM.

I guess there is more lack of documentation issue (or maybe absent of kdoc) then API inconsistency. 
>  And I already get this behavior with every database, but not with H2.

AFAIK some jdbc-drivers pre-fetch some additional rows from database (sometimes size is configurated via parameters), but i think that if you run same code on large dataset then you will achieve similar exception on other databases (I will check it tomorrow).

>  There is also a bug (?) with SQLite: it doesn't support current default transaction isolation level. 

This strange, I will check it too.

**Comment by bluemix** [comment]
any updates?

**Comment by dya-tel** [comment]
Well, it seems that nested queries do not work anywhere and .toList() is a good enough solution when this behaviour is required.

**Comment by dillius** [comment]
@dya-tel I'm trying to get around the default isolation level issue as well but no point of change seems to make it accept a different default isolation level for Sqlite. Could you share some insight as to how you altered it to make it accept it?

**Comment by Tapac** [comment]
@dillius , set needed transaction isolation level after Database.connect:
```
TransactionManager.current.defaultIsolationLevel = Connection.TRANSACTION_READ_COMMITTED
```
or within connect:
```
Database.connect(datasource) { ThreadLocalTransactionManager(it, Connection.TRANSACTION_READ_COMMITTED) }
```

**Comment by Tapac** [comment]
Happy news, everyone!
Looks like we found workaround for that issue and now you can work with h2 without making toList() on queries.
Exposed will fetch whole result set on getting iterator.
@dya-tel , @bluemix 
Will be available on next release (0.8.8 or 0.9).

**Comment by dya-tel** [comment]
Nice!
By the way, what about the default isolation level, which is unsupported by SQLite?
